akka-docs/src/main/paradox/scala/guide/tutorial_5.md

```diff
@@ -0,0 +1,253 @@
+   * We can reach the deadline and receive a `CollectionTimeout`.
+
+In the first two cases, we need to keep track of the replies, which we now simply delegate to a method `receivedResponse`, which we will discuss later. In the case of timeout, we need to simply take all the actors that have not yet replied yet (the members of the set `stillWaiting`) and put a `DeviceTimedOut` as the status in the final reply. Then we reply to the submitter of the query with the collected results and stop the query actor.
+
+To accomplish this, add the following to your `DeviceGroupQuery` source file:
+
+
+
+Scala
+:   @@snip [DeviceGroupQuery.scala]($code$/scala/tutorial_5/DeviceGroupQuery.scala) { #query-state }
+
+Java
+:   @@snip [DeviceGroupQuery.java]($code$/java/jdocs/tutorial_5/DeviceGroupQuery.java) { #query-state }
+
+It is not yet clear how we will "mutate" the `answersSoFar` and `stillWaiting` data structures. One important thing to note is that the function `waitingForReplies` **does not handle the messages directly. It returns a `Receive` function that will handle the messages**. This means that if we call `waitingForReplies` again, with different parameters,
+then it returns a brand new `Receive` that will use those new parameters.
+
+We have seen how we
+can install the initial `Receive` by simply returning it from `receive`. In order to install a new one, to record a
+new reply, for example, we need some mechanism. This mechanism is the method `context.become(newReceive)` which will
+_change_ the actor's message handling function to the provided `newReceive` function. You can imagine that before
+starting, your actor automatically calls `context.become(receive)`, i.e. installing the `Receive` function that
+is returned from `receive`. This is another important observation: **it is not `receive` that handles the messages,
+it just returns a `Receive` function that will actually handle the messages**.
+
+We now have to figure out what to do in `receivedResponse`. First, we need to record the new result in the map `repliesSoFar` and remove the actor from `stillWaiting`. The next step is to check if there are any remaining actors we are waiting for. If there is none, we send the result of the query to the original requester and stop the query actor. Otherwise, we need to update the `repliesSoFar` and `stillWaiting` structures and wait for more
+messages.
+
+In the code before, we treated `Terminated` as the implicit response `DeviceNotAvailable`, so `receivedResponse` does
+not need to do anything special. However, there is one small task we still need to do. It is possible that we receive a proper
+response from a device actor, but then it stops during the lifetime of the query. We don't want this second event
+to overwrite the already received reply. In other words, we don't want to receive `Terminated` after we recorded the
+response. This is simple to achieve by calling `context.unwatch(ref)`. This method also ensures that we don't
```

Comment:
Is this correct? Isn't it possible for a device actor to send Device.RespondTemperature message and then terminate just after that? If DeviceGroupQuery actor is busy handling other message then both Device.RespondTemperature and Terminated message will be added to its inbox. Will context.unwatch remove Terminated message from the inbox? If not then DeviceGroupQuery will first handle Device.RespondTemperature message and then Terminated message for the same device. This would lead to a good temperature reading being overridden with DeviceGroup.DeviceNotAvailable.

Reply:
I think you're right there's still a window there. Would you want to propose an alternate wording?

Reply:
Actually I may be wrong. I've just noticed the next sentence "This method also ensures that we don't receive Terminated events that are already in the mailbox of the actor." so it should all be fine I guess. If that sentence is correct then Terminated message will be removed from the inbox by the unwatch call. Sorry for making a fuss.